aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -0,0 +1,514 @@
+    for (size_t i = 0; i < perm_.size(); i++) {
+      res[i] = data[perm_[i]];
+    }
+    return res;
+  };
+
+  // Update shape and strides
+  shape_ = reorder(shape_, perm_);
+  for (auto& op : operands_) {
+    if (op.stride_.size() > 0) {
+      op.stride_ = reorder(op.stride_, perm_);
+    }
+  }
+}
+
+template <typename F>
+static std::tuple<ScalarType, Backend>
+compute_result_type(at::ArrayRef<OperandInfo> operands, const F& predicate) {
+  auto result_type = ScalarType::Undefined;
+  auto backend = Backend::Undefined;
+  for (auto& op : operands) {
+    if (!op.tensor_->defined()) continue;
+    if (!predicate(*op.tensor_)) continue;
+    auto dtype = op.tensor_->type().scalarType();;
+    result_type = (result_type == ScalarType::Undefined
+        ? dtype
+        : promoteTypes(result_type, dtype));
+    backend = (backend == Backend::Undefined
+        ? op.tensor_->type().backend()
+        : backend);
+  }
+  return std::make_tuple(result_type, backend);
+}
+
+void TensorIterator::compute_common_type() {
+  // The result dtype is computed with the precedence:
+  // 1) Tensors of rank one or higher
+  // 2) Tensors of any rank that aren't wrapped numbers
+  // 3) Any tensor
+  //
+  // For example:
+  //  torch.randn(3, 3) + torch.tensor([1.0], dtype=torch.double) -> torch.double
+  //  torch.randn(3, 3) + torch.tensor(1.0, dtype=torch.double) -> torch.float
+  //  torch.tensor(3, dtype=torch.uint8) + 5 -> torch.uint8
+  //
+  auto result_type = ScalarType::Undefined;
+  auto backend = Backend::Undefined;
+  std::tie(result_type, backend) = compute_result_type(operands_, [](const Tensor& t) {
+    return t.dim() > 0;
+  });
+  if (result_type == ScalarType::Undefined) {
+    std::tie(result_type, backend) = compute_result_type(operands_, [](const Tensor& t) {
+      return !t.get()->is_wrapped_number();
+    });
+  }
+  if (result_type == ScalarType::Undefined) {
+    std::tie(result_type, backend) = compute_result_type(operands_, [](const Tensor& t) {
+      return true;
+    });
+  }
+
+  AT_ASSERT(result_type != ScalarType::Undefined);
+  AT_ASSERT(backend != Backend::Undefined);
+
+  auto& type = at::globalContext().getType(backend, result_type);
+
+  for (auto& op : operands_) {
+    if (!op.type_) {
+      op.type_ = &type;
+      if (op.tensor_->defined() && type != op.tensor_->type()) {
+        if (op.tensor_->dim() == 0) {
+          if (type.backend() != at::kCUDA) {
+            *op.tensor_ = op.tensor_->toType(type);
+          }
+        } else {
+          op.needs_cast_ = true;
+        }
+      }
+    }
+  }
+}
+
+DimVector TensorIterator::compatible_stride(int element_size) const {
+  auto stride = DimVector();
+  stride.push_back(element_size);
+  for (int i = 0; i < ndim() - 1; i++) {
+    stride.push_back(shape_[i] * stride[i]);
+  }
+  return stride;
+}
+
+DimVector TensorIterator::invert_perm(IntList input) const {
+  // Invert the permutation caused by reorder_dimensions. This is not valid
```

Comment:
might want to just record that coalesce_dimensions was called and assert it wasn't called here.